core/stats_handler.go

```diff
@@ -0,0 +1,203 @@
+package core
+
+import (
+	"context"
+	"time"
+
+	tpb "github.com/textileio/go-threads/api/pb"
+	hpb "github.com/textileio/textile/v2/api/hubd/pb"
+	mdb "github.com/textileio/textile/v2/mongodb"
+	"google.golang.org/grpc/stats"
+)
+
+type StatsHandler struct {
+	t *Textile
+}
+
+var _ stats.Handler = (*StatsHandler)(nil)
+
+var statsTimeout = time.Second * 10
+
+// HandleRPC accounts for customer usage across services.
+func (h *StatsHandler) HandleRPC(ctx context.Context, st stats.RPCStats) {
+	switch st := st.(type) {
+	case *stats.OutPayload:
+		if getStats(ctx) == nil {
+			return
+		}
+
+		// Handle payload types.
+		egress := int64(st.WireLength)
+		var reads, writes int64
+		var pl interface{}
+		switch spl := st.Payload.(type) {
+		case *tpb.ReadTransactionReply:
+			pl = spl.Option
+		case *tpb.WriteTransactionReply:
+			pl = spl.Option
+		default:
+			pl = spl
+		}
+		switch pl := pl.(type) {
+
+		// Don't charge for usage when customer is trying to add a payment method, cancel subscription, etc.
+		case *hpb.GetBillingSessionResponse:
+			egress = 0
+
+		// Account for threaddb reads and writes
+		case *tpb.CreateReply:
+			if pl.InstanceIDs != nil {
+				writes = int64(len(pl.InstanceIDs))
+			}
+		case *tpb.WriteTransactionReply_CreateReply:
+			if pl.CreateReply.TransactionError == "" {
+				writes = 1
+			}
+		case *tpb.VerifyReply:
+			if pl.TransactionError == "" {
+				reads = 1
+			}
+		case *tpb.WriteTransactionReply_VerifyReply:
+			if pl.VerifyReply.TransactionError == "" {
+				reads = 1
+			}
+		case *tpb.SaveReply:
+			if pl.TransactionError == "" {
+				writes = 1
+			}
+		case *tpb.WriteTransactionReply_SaveReply:
+			if pl.SaveReply.TransactionError == "" {
+				writes = 1
+			}
+		case *tpb.DeleteReply:
+			if pl.TransactionError == "" {
+				writes = 1
+			}
+		case *tpb.WriteTransactionReply_DeleteReply:
+			if pl.DeleteReply.TransactionError == "" {
+				writes = 1
+			}
+		case *tpb.HasReply:
```

Comment:
Honest question: Is there a reason you don't just "stack" all these cases and fall through? Is it just because the transaction errors are typed like this?

Reply:
Here, `pl` is `interface{}` unless it's in one of the switch statements... so, each one is needed in order to pull params out of the payloads.